libcache/cache.c

```diff
@@ -0,0 +1,650 @@
+
+		/* write-back */
+		if (IS_DIRTY(temp->flags)) {
+			written = cache_flushLine(writeCb, temp, offset, count);
+			if (written == -1) {
+				return -1;
+			}
+			free(temp->data);
+		}
+
+		for (i = 0; i < set->count; ++i) {
+			if (IS_VALID(set->lines[i].flags) && (set->lines[i].tag == temp->tag)) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+	}
+
+	written = cache_executePolicy(writeCb, temp, offset, count, policy);
+
+	qsort(set->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	return written;
+}
+
+
+static cacheline_t *cache_findLine(cacheset_t *set, uint64_t tag)
+{
+	cacheline_t key, *keyPtr = &key, **linePtr = NULL;
+
+	key.tag = tag;
+
+	linePtr = bsearch(&keyPtr, set->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	return (linePtr != NULL) ? *linePtr : NULL;
+}
+
+
+static cacheline_t *cache_updateTimestamps(cacheset_t *set, uint64_t tag)
+{
+	cacheline_t *linePtr = NULL;
+
+	linePtr = cache_findLine(set, tag);
+
+	if (linePtr != NULL) {
+		LIST_REMOVE(&set->timestamps, linePtr);
+		LIST_ADD(&set->timestamps, linePtr);
+
+		return linePtr;
+	}
+
+	return NULL;
+}
+
+
+static void *cache_readFromSet(cacheset_t *set, uint64_t tag, uint64_t offset)
+{
+	cacheline_t *linePtr = NULL;
+
+	linePtr = cache_updateTimestamps(set, tag);
+
+	return (linePtr != NULL) ? (unsigned char *)(linePtr)->data + offset : NULL;
+}
+
+
+static uint64_t cache_compOffset(const cachectx_t *cache, const uint64_t addr)
+{
+	return addr & cache->offMask;
+}
+
+
+static uint64_t cache_compSet(const cachectx_t *cache, const uint64_t addr)
+{
+	return (addr >> cache->offBitsNum) & cache->setMask;
+}
+
+
+static uint64_t cache_compTag(const cachectx_t *cache, const uint64_t addr)
+{
+	return (addr >> (cache->offBitsNum + cache->setBitsNum)) & cache->tagMask;
+}
+
+
+/* TODO: add error handling */
+ssize_t cache_write(cachectx_t *cache, uint64_t addr, void *buffer, size_t count, int policy)
+{
+	ssize_t position = 0, ret = 0;
+	void *data = NULL, *dest = NULL;
+	unsigned char flags = 0;
+	uint64_t index = 0, offset = 0, tag = 0;
+	size_t pieceSize, left = 0, remainder = 0;
+	cacheline_t line, *linePtr;
+
+	if (cache == NULL || buffer == NULL) {
+		return -1;
+	}
+
+	if (count == 0 || (policy != LIBCACHE_WRITE_BACK && policy != LIBCACHE_WRITE_THROUGH)) {
+		return 0;
+	}
+
+	offset = cache_compOffset(cache, addr);
+
+	left = count;
+
+	while (1) {
+		index = cache_compSet(cache, addr);
+		tag = cache_compTag(cache, addr);
+
+		if (left == count) {
+			if (count <= (cache->lineSize - offset)) {
+				pieceSize = count;
+			}
+			else {
+				pieceSize = cache->lineSize - offset;
+				remainder = (left - pieceSize) % cache->lineSize;
+			}
+			addr -= offset;
+		}
+		else if (left == remainder) {
+			pieceSize = remainder;
+			offset = 0;
+		}
+		else {
+			pieceSize = cache->lineSize;
+			offset = 0;
+		}
+
+		data = cache_readFromSet(&cache->sets[index], tag, offset);
```

Comment:
It would be better to findLine here - if data != NULL we'll do that anyway, so there would be 2 seaches in total.

Reply:
Solved, thank you